File: app/validation/Unique.java
package validation;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import net.sf.oval.configuration.annotation.Constraint;

/**
 * Checks that the field is unique in the database
 */
@Retention(RetentionPolicy.RUNTIME)
@Target( { ElementType.FIELD, ElementType.PARAMETER })
@Constraint(checkWith = UniqueCheck.class)
public @interface Unique {
    //String message() default UniqueCheck.mes;
    String message() default "validation.unique";
}

File: app/validation/unique.scala
package validation

import play.exceptions.UnexpectedException
import net.sf.oval.context.FieldContext
import net.sf.oval.Validator
import net.sf.oval.configuration.annotation.AbstractAnnotationCheck
import net.sf.oval.context.OValContext

import play.db.jpa._

class UniqueCheck extends AbstractAnnotationCheck[Unique] {
    override def isSatisfied(validatedObject: Object, value: Object, context: OValContext, validator: Validator) = {
        if(!validatedObject.isInstanceOf[play.db.jpa.Model] || !context.isInstanceOf[FieldContext]) {
            throw new UnexpectedException("@Unique annotation can only be applied to fields of a Model object")
        }
        
        val model = validatedObject.asInstanceOf[play.db.jpa.Model]
        val fieldName = context.asInstanceOf[FieldContext].getField().getName
        val count = JPQL.instance.count(validatedObject.getClass().getName, fieldName + " = ?", Array(value))
        count == 0
    }
}
